Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,191 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A base ByteBuffer "Slicer", this base one just splits by a separator.
+public struct ByteBufferSliceSplitIterator: IteratorProtocol {
+    
+    var byteBuffer: ByteBuffer
+    var separator: UInt8
+    let start: Int
+    let length: Int
+    
+    private var currentIndex: Int
+    
+    public typealias Element = ByteBuffer
+    mutating public func next() -> ByteBuffer? {
+        
+        return byteBuffer.withVeryUnsafeBytes { pointer -> ByteBuffer? in
+            
+            let startPoint   = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            
+            var address      = startPoint.advanced(by: currentIndex)
+            let finalAddress =    address.advanced(by: length)
+            
+            let initialPointer = address
+
+            while address < finalAddress && address.pointee != 0 {
+                
+                if address.pointee == separator {
+                    break
+                }
+                address = address.advanced(by: 1)
+            }
+            
+            let tokenLength = address - initialPointer
+            let ret = byteBuffer.getSlice(at: currentIndex, length: tokenLength)
```

Comment:
I think it would be better to return the slice index and token length from the call to `withVeryUnsafeBytes` and then emit the slice, rather than build the slice in here. That resolves the possibility that `withVeryUnsafeBytes` may end up marked as a mutating method at some time.